src/Microsoft.Data.Analysis/DataFrame.IO.cs

```diff
@@ -306,5 +309,112 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
+        /// <param name="separator">column separator</param>
+        /// <param name="header">has a header or not</param>
+        /// <param name="encoding">The character encoding. Defaults to UTF8 if not specified</param>
+        /// <param name="cultureInfo">culture info for handeling decimal values</param>
+        public static void WriteCsv(DataFrame dataFrame, string path,
+                                   char separator = ',', bool header = true,
+                                   Encoding encoding = null, CultureInfo cultureInfo = null)
+        {
+            using (FileStream csvStream = new FileStream(path, FileMode.Create))
+            {
+                WriteCsv(dataFrame: dataFrame, csvStream: csvStream,
+                           separator: separator, header: header,
+                           encoding: encoding, cultureInfo: cultureInfo);
+            }
+        }
+
+        /// <summary>
+        /// Writes a DataFrame into a CSV.
+        /// Follows pandas API.
+        /// </summary>
+        /// <param name="dataFrame"><see cref="DataFrame"/></param>
+        /// <param name="csvStream">stream of CSV data to be write out</param>
+        /// <param name="separator">column separator</param>
+        /// <param name="header">has a header or not</param>
+        /// <param name="encoding">the character encoding. Defaults to UTF8 if not specified</param>
+        /// <param name="cultureInfo">culture info for handeling decimal values</param>
+        public static void WriteCsv(DataFrame dataFrame, Stream csvStream,
+                           char separator = ',', bool header = true,
+                           Encoding encoding = null, CultureInfo cultureInfo = null)
+        {
+            if (cultureInfo is null)
+            {
+                cultureInfo = CultureInfo.InvariantCulture;
+            }
+            else
+            {
+                if (cultureInfo.NumberFormat.NumberDecimalSeparator.Equals(",") && separator.Equals(','))
+                {
+                    throw new ArgumentException("Decimal separator cannot match the column separator");
+                }
+            }
+
+            using (StreamWriter csvFile = new StreamWriter(csvStream, encoding ?? Encoding.UTF8))
+            {
+                if (dataFrame != null)
+                {
+                    var columnNames = dataFrame.Columns.GetColumnNames();
+
+                    if (header)
+                    {
+                        var headerColumns = string.Join(separator.ToString(), columnNames);
+                        csvFile.WriteLine(headerColumns);
+                    }
+
+                    var currPosition = csvStream.Position;
+
+                    foreach (var row in dataFrame.Rows)
+                    {
+                        var record = new StringBuilder();
+
+                        bool firstRow = true;
+                        foreach (var cell in row)
```

Comment:
(minor perf question/suggestion): I wonder if we used the `IDataView` API here we could avoid the boxing and unboxing of every primitive value in the DataFrame.

For example, for each column, we can get a `Getter` and use it to get the value generically and write it to the StringBuilder.

See [https://github.com/dotnet/machinelearning/blob/5370692f6d1e6f1f686596d530032b4226dc9e80/src/Microsoft.ML.Data/DataDebuggerPreview.cs#L44-L60](https://github.com/dotnet/machinelearning/blob/5370692f6d1e6f1f686596d530032b4226dc9e80/src/Microsoft.ML.Data/DataDebuggerPreview.cs#L44-L60) for a general structure on how to do this.

I think what we have here is good for a first implementation to get the functionality added. But we may want to benchmark alternatives to see how well this performs.